File: frontend/src/domain/entities/Product.ts
import type { Name } from '../valuesobject/Name';
import type { Price } from '../valuesobject/Price';
import type { Quantity } from '../valuesobject/Quantity';

export class Product {
  private constructor(private price: Price, private quantity: Quantity, private name: Name) {}

  public static create(price: Price, quantity: Quantity, name: Name): Product {
    return new Product(price, quantity, name);
  }

  public getProductForRegister() {
    return {
      name: this.name.getValue(),
      price: this.price.getValue(),
      quantity: this.quantity.getValue(),
    };
  }
}

File: backend/src/domain/entities/Product.go
package entities

import "github.com/marcosfrancomarinho/go-first-project/src/domain/valuesobject"

type Product struct {
	id       *valuesobject.ID
	name     *valuesobject.Name
	price    *valuesobject.Price
	quantity *valuesobject.Quantity
}

func NewProduct(
	id *valuesobject.ID,
	name *valuesobject.Name,
	price *valuesobject.Price,
	quantity *valuesobject.Quantity,
) *Product {
	return &Product{id, name, price, quantity}
}

func (p *Product) GetName() string {
	return p.name.GetValue()
}
func (p *Product) GetID() string {
	return p.id.GetValue()
}

func (p *Product) GetPrice() float64 {
	return p.price.GetValue()
}
func (p *Product) GetQuantity() int {
	return p.quantity.GetValue()
}

func (p *Product) GetTotal() float64 {
	return p.GetPrice() * float64(p.GetQuantity())
}
